### PR TITLE
Reference system

### PR DESCRIPTION
**Pull request by JoshBroomberg**

### CODE REVIEW

File: app/models/link_click.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
+class LinkClick < ActiveType::Object
+  attr_accessor :user
+  attr_accessor :link
+  attr_accessor :cookie_data
+
+  validates_presence_of :link
+
+  after_save :generate_new_cookie_data, unless: :register_click
+
+  def register_click
```
Comment (JoshBroomberg) [inline app/models/link_click.rb]: `click_registered?`

File: app/models/link_click.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    return unless user
```
Comment (JoshBroomberg) [inline app/models/link_click.rb]: `false`

File: app/models/link_click.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    Click.create!(user_id: user.id, link_id: link.id)
```
Comment (JoshBroomberg) [inline app/models/link_click.rb]: remove `!`

more explicitly return a boolean `!!Click.create(...)`

File: app/models/link_click.rb [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+  end
+
+  def generate_new_cookie_data
```
Comment (JoshBroomberg) [inline app/models/link_click.rb]: remove `_data` on everythign

File: app/controllers/reference/link_clicks_controller.rb [hunk 5]
```diff
@@ -0,0 +1,25 @@
+class Reference::LinkClicksController < ApplicationController
+  before_action :load_link
+  before_action :load_cookie_data
+
+  def create
+    link_click = LinkClick.new(link: @link, user: current_user, cookie_data: @cookie_data)
+    if link_click.save
+      write_cookie(:link_data, link_click.cookie_data.to_json)
```
Comment (JoshBroomberg) [inline app/controllers/reference/link_clicks_controller.rb]: `clicked_links`

File: app/controllers/reference/link_clicks_controller.rb [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    else
+      raise_page_not_found
+    end
+  end
+
+  private
+
+  def load_link
+    @link ||= Link.find_by(link_path: params[:link_path])
+    raise_page_not_found unless @link
```
Comment (JoshBroomberg) [inline app/controllers/reference/link_clicks_controller.rb]: Should this raise an error/404 or just redirect to ixp.co.za/? Issue witht he redirect is that if people type the code wrong and just get redirected, they won't know they typed it wrong and won't be tracked...

**Comment by JoshBroomberg** [comment]
@khanotations this is ready to review. I cleaned things up quite a bit and write tests for the LinkClick class (which is the key logic class). 

- Still no UI for this. The new, more flexible model structure adds some complexity to the UI which we will need to talk about. 
- I elected not to do controller tests because the actions in the controllers are pretty simple, and integration tests would probably be the easiest way to test this stuff. Analogous to the rest of the system which doesn't yet have, but probably does need, integration tests.

File: app/controllers/reference/link_clicks_controller.rb (hunk 6 above)
Comment (khanotations) [inline app/controllers/reference/link_clicks_controller.rb]: I think we should have a special "Link not found" page

File: app/controllers/reference/link_clicks_controller.rb [hunk 7]
```diff
@@ -0,0 +1,25 @@
+class Reference::LinkClicksController < ApplicationController
+  before_action :load_link
+  before_action :load_cookie
+
+  def create
+    link_click = LinkClick.new(link_id: @link.id, user_id: current_user.try(:id), cookie: @cookie)
+    if link_click.save
+      write_cookie(:links, link_click.cookie.to_json)
+      redirect_to @link.redirect_url
```
Comment (khanotations) [inline app/controllers/reference/link_clicks_controller.rb]: Do you think we should keep URL parameters? like it's annoying to have to set up a different link for http://a.com?a=b and http://a.com?c=d, when we could do http://ixp.co.za/link?a=b. ya feel me

File: app/controllers/application_controller.rb [hunk 8]
```diff
@@ -13,6 +13,18 @@ def render_json_error(msg, status = :unprocessable_entity)
     render json: msg, status: status
   end
 
+  def raise_page_not_found
```
Comment (khanotations) [inline app/controllers/application_controller.rb]: I think this should be a custom page for bad references.

File: app/views/layouts/application.html.erb [hunk 9]
```diff
@@ -18,7 +18,6 @@
   <!-- End wrapper-->
   <%= javascript_include_tag 'application' %>
   <%= javascript_include_tag 'student' %>
-  <%= intercom_script_tag %>
```
Comment (khanotations) [inline app/views/layouts/application.html.erb]: yay

File: app/models/user.rb [hunk 10]
```diff
@@ -8,11 +8,14 @@ class User < ActiveRecord::Base
...
 
-  # This relation should not be used directly. Rather use relations on user subclasses.
+  # These relations should not be used directly. Rather use relations on user subclasses.
   has_many :responses, dependent: :destroy
+  has_many :clicks, dependent: :destroy
+  has_many :references, dependent: :destroy
   # Intent relations.
   # See "customising queries" here: http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html
   has_many :intents, dependent: :destroy
+
```
Comment (khanotations) [inline app/models/user.rb]: remove pls

**Pull request merged**